src/Components/Web/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IVirtualizeJsCallbacks, IAsyncDisposable
+    {
+        private VirtualizeJsInterop _jsInterop = default!;
+
+        private ElementReference _spacerBefore;
+
+        private ElementReference _spacerAfter;
+
+        [Inject]
+        private IJSRuntime JSRuntime { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the number of items rendered before the visible part of the container.
+        /// </summary>
+        protected int ItemsBefore { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items the container can visibly fit.
+        /// </summary>
+        protected int VisibleItemCapacity { get; private set; }
+
+        /// <summary>
+        /// Gets the number of items currently visible within the container.
+        /// </summary>
+        protected int ItemsVisible { get; private set; }
```

Comment:
This might be nitpicky, but is this property required? It seems like it could be computed from the others, i.e.,

ItemsVisible = Math.Min(ItemCount - ItemsBefore, VisibleItemCapacity)

Reply:
Was the goal behind this to reduce the public API surface? I've combined `VirtualizeDeferred` and `VirtualizeFixed` and removed the need to make it a protected property.